Bugzilla/Config.pm

```diff
@@ -11,375 +11,366 @@ use 5.10.1;
+  }
+  else {
+    try {
+      my $dbh  = Bugzilla->dbh;
+      my $rows = $dbh->selectall_arrayref('SELECT name, value FROM params');
+      foreach my $row (@$rows) {
+        my ($name, $value) = @$row;
+        $self->{params}->{$name} = $value;
+      }
+    }
+    catch {
+      WARN("Database not yet available: $_");
 
-sub param_panels {
-  my $param_panels = {};
-  my $libpath      = bz_locations()->{'libpath'};
-  foreach my $item ((glob "$libpath/Bugzilla/Config/*.pm")) {
-    $item =~ m#/([^/]+)\.pm$#;
-    my $module = $1;
-    $param_panels->{$module} = "Bugzilla::Config::$module"
-      unless $module eq 'Common';
+      # Load defaults if database is unavailable
+      my $defs = $self->_load_defs();
+      foreach my $key (keys %{$defs}) {
+        $self->{params}->{$key} = $defs->{$key}->{default};
+      }
+    };
   }
 
-  # Now check for any hooked params
-  Bugzilla::Hook::process('config_add_panels', {panel_modules => $param_panels});
-  return $param_panels;
-}
+  $cache->{params_obj} = $self;
 
-sub SetParam {
-  my ($name, $value) = @_;
+  return $self;
+}
 
-  _load_params unless %params;
-  die "Unknown param $name" unless (exists $params{$name});
+###############################
+####       Setters         ####
+###############################
 
-  my $entry = $params{$name};
+sub set_param {
+  my ($self, $name, $value) = @_;
 
-  # sanity check the value
+  die "Unknown param $name" unless (exists $self->{params}->{$name});
 
+  # Sanity check the value
   # XXX - This runs the checks. Which would be good, except that
   # check_shadowdb creates the database as a side effect, and so the
   # checker fails the second time around...
-  if ($name ne 'shadowdb' && exists $entry->{'checker'}) {
-    my $err = $entry->{'checker'}->($value, $entry);
+  my $def = $self->{param_defs}->{$name};
+  if ($name ne 'shadowdb' && exists $def->{'checker'}) {
+    my $err = $def->{'checker'}->($value, $def);
     die "Param $name is not valid: $err" unless $err eq '';
   }
 
-  Bugzilla->params->{$name} = $value;
+  $self->{params}->{$name} = $value;
 }
 
-sub update_params {
-  my ($params) = @_;
-  my $answer = Bugzilla->installation_answers;
+sub update {
+  my ($self, $params) = @_;
+  $params ||= $self->{params};
 
-  my $param = read_param_file();
-  my %new_params;
+  try {
+    my $dbh = Bugzilla->dbh;
+    foreach my $key (keys %{$params}) {
+      my $new_value = $params->{$key} || '';
+      if ($dbh->selectrow_array('SELECT 1 FROM params WHERE name = ?', undef, $key)) {
+        $dbh->do('UPDATE params SET value = ? WHERE name = ?', undef, $new_value, $key);
+      }
+      else {
+        $dbh->do('INSERT INTO params (name, value) VALUES (?, ?)',
+          undef, $key, $new_value);
+      }
+    }
+
+    # And now we have to reset the params cache
+    Bugzilla->memcached->set_params($params);
+    delete Bugzilla->request_cache->{params};
+    delete Bugzilla->request_cache->{params_obj};
```

Comment:
Do you know why we have two caches - one in `memcache`, and one here in-memory (if I understand correctly)?
Perhaps we need only one local cache: I'm guessing the in-memory `Bugzilla->request_cache` one.

Reply:
memcache is a cached of just the hash of parameter name and values. It can be used instead of reading from the params table of the database.

In memory is the request_cache that is destroyed at the end of each client request.